File: runelite-client/src/main/java/net/unethicalite/api/script/blocking_events/BlockingEventManager.java
package net.unethicalite.api.script.blocking_events;

import lombok.Getter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

@Getter
public class BlockingEventManager
{
	private final Map<LoginEvent.Response, Function<LoginEvent.Response, Integer>> loginMessageHandlers =
			new HashMap<>();
	private final List<BlockingEvent> blockingEvents = new ArrayList<>();

	public BlockingEventManager()
	{
		blockingEvents.add(new LoginEvent(this));
		blockingEvents.add(new WelcomeScreenEvent());
		blockingEvents.add(new ResizableEvent());
		blockingEvents.add(new DeathEvent());
		blockingEvents.add(new BankTutorialEvent());
	}

	public void addLoginResponseHandler(LoginEvent.Response response, Function<LoginEvent.Response, Integer> handler)
	{
		loginMessageHandlers.put(response, handler);
	}

	public void removeLoginResponseHandler(LoginEvent.Response response)
	{
		loginMessageHandlers.remove(response);
	}

	public LoginEvent getLoginEvent()
	{
		return (LoginEvent) blockingEvents.stream().filter(e -> e instanceof LoginEvent)
				.findFirst()
				.orElse(null);
	}

	public boolean remove(Class<? extends BlockingEvent> clazz)
	{
		return blockingEvents.removeIf(e -> e.getClass().equals(clazz));
	}

	public boolean add(BlockingEvent event)
	{
		return blockingEvents.add(event);
	}
}

File: runelite-client/src/main/java/dev/unethicalite/client/script/blocking_events/BlockingEventManager.kt
package dev.unethicalite.client.script.blocking_events

class BlockingEventManager {

    companion object {
        val loginMessageHandlers: MutableMap<LoginEvent.Response, (LoginEvent.Response) -> Int> = mutableMapOf()
    }

    val blockingEvents = mutableSetOf(LoginEvent(), WelcomeScreenEvent(), ResizableEvent(), DeathEvent())

    fun addLoginResponseHandler(vararg responses: LoginEvent.Response, handler: (LoginEvent.Response) -> Int) {
        for (response in responses) {
            loginMessageHandlers[response] = handler
        }
    }

    fun removeLoginResponseHandler(message: LoginEvent.Response) {
        loginMessageHandlers.remove(message)
    }

    fun getLoginEvent(): LoginEvent? {
        return blockingEvents.firstOrNull { it.javaClass == LoginEvent::class.java } as LoginEvent?
    }

    /**
     * Removes a blocking event matching given class.
     */
    fun <T : BlockingEvent> remove(clazz: Class<T>) {
        blockingEvents.removeIf { it.javaClass == clazz }
    }

    /**
     * Gets a blocking event in the list matching given class.
     */
    @Suppress("UNCHECKED_CAST")
    fun <T : BlockingEvent> get(clazz: Class<T>): T? {
        if (blockingEvents.isEmpty()) {
            return null
        }

        return blockingEvents.firstOrNull { it.javaClass == clazz } as T
    }

    /**
     * Adds a new blocking event.
     */
    fun add(event: BlockingEvent) {
        blockingEvents.add(event)
    }
}
